docs: add threat model section (#4697)
cilium/tetragon: docs/content/en/docs/threat-model/_index.md

**Recommended controls:**

- Protection against network attackers requires implementing encryption in transit for all data exported from Tetragon. Security event collection should use HTTPS or other encrypted protocols when transmitting data to external storage, SIEM platforms, or analysis systems. The specific implementation depends on the chosen event export mechanism, but all network-based exports should use strong encryption with proper certificate validation to prevent man-in-the-middle attacks.
- Prometheus metrics collection also benefits from encryption. Enabling TLS on the Prometheus metrics endpoint prevents casual interception of operational metrics, and configuring Prometheus scrapers to use authenticated TLS connections ensures both confidentiality and authenticity of the metrics channel. Users should weigh the operational overhead of managing TLS certificates for metrics collection against the risks associated with metrics disclosure in their specific threat model.

> ## Operational Security Considerations

Reviewer: I'm not certain where the best place would be to add this, but a frequent report we receive to the OSS mailing list concerns TOCTOU issues due to hooking on syscalls. It would be good to link to the warning at the top of https://tetragon.io/docs/concepts/tracing-policy/hooks/, to highlight to users that there is a security/visibility risk in using the wrong hook points when writing policies, and that this does not indicate a security vulnerability in Tetragon itself.

- Tetragon generates events that can be incorporated into existing incident response workflows, SIEM platforms, and security analytics tools. The volume, format, and semantic content of these events influence the success of integration. Users should implement appropriate event filtering, aggregation, and enrichment to ensure Tetragon data is actionable within their operational context. Excessive false positives can lead to alert fatigue and undermine security effectiveness, highlighting the importance of tuning and validating detection rules.

## Conclusion

While eBPF provides unparalleled visibility and enforcement capabilities, its security is still tied to the integrity of  the adjacent environment it operates in, particularly the host operating system and the Kubernetes control plane. Implementing mitigation strategies that encompass the recommended controls in this threat model will ensure Tetragon continues to remain a trusted source of truth. The table below summarizes the attacker profiles, primary vectors, severity, and the primary mitigation.

> 

Reviewer: I don't think the table below is how we should be summarising our recommendations. A lot of the suggestions above are specific to particular use cases, and the table below implies that reducing risk is a matter of implementing the primary mitigation. I think it makes more sense not to have the table - operators should read the entire set of recommendations and assess which potential controls are relevant/useful for them.

deployment scenarios, including Kubernetes-based deployments using DaemonSets,
standalone containerized agents, and systemd-managed service deployments on
individual hosts.

{{< figure src="/images/smart_observability.png"
caption="Tetragon Overview Diagram"
width=800px
alt="A diagram showing Tetragon capabilities"
>}}

This analysis evaluates threats associated with five distinct attacker profiles,
each representing varying levels of access and privilege within the system.
These profiles encompass attackers operating within Kubernetes workloads,
individuals with limited host privileges, attackers with root-equivalent access,
attackers who have compromised the Kubernetes API server, and network-based
attackers capable of intercepting or manipulating network traffic.  The threat
model assumes Tetragon operates on fully patched systems with appropriate
security hardening, including [Kubernetes security best
practices](https://kubernetes.io/docs/tasks/administer-cluster/securing-a-cluster/)
where applicable. This baseline assumption enables a focus on threats specific
to Tetragon’s design and operation, rather than general system vulnerabilities.

> **N/B:** There is a security/visibility risk in using the wrong hook points when

Author: @ferozsalam added the caveat here. 

I'll add another one to the tracing policy section in seperate PR since the change is unrelated to this threat model

writing tetragon tracing policies, this however does not indicate a security
vulnerability in Tetragon itself. Operators are advised to ensure that chosen
hook points are validated carefully to achieve the desired security/visibility

> objective.

Reviewer 2: You can choose between note/caution/warning. Usually this would be something else but I feel like note is okay because it's in the context of the threat model relation to this.
```suggestion
{{< note >}}
There is a security/visibility risk in using the wrong hook points when
writing tetragon tracing policies, this however does not indicate a security
vulnerability in Tetragon itself. Operators are advised to ensure that chosen
hook points are validated carefully to achieve the desired security/visibility
objective.
{{< /note >}}
```

pipeline, leading to event loss, performance degradation, or storage exhaustion.
This denial-of-service attack vector exploits Tetragon's core requirement to
observe and record security-relevant activities, thereby turning a fundamental
strength into a potential vulnerability.

Current analysis indicates that such attackers cannot directly compromise
Tetragon's integrity or availability beyond denial-of-service scenarios. The
isolation enforced by container boundaries, together with Tetragon's operation
at the kernel level, prevents workload-level attackers from manipulating
Tetragon's eBPF programs, configuration, or core functionality. Nevertheless,
the denial-of-service vector remains a significant concern that requires
operational controls.

**Recommended controls**:

Proactive monitoring can mitigate workload-based denial-of-service attacks.
Operators are encouraged to monitor and alert on dropped events via [Tetragon
Health Metrics]({{< ref "/docs/installation/metrics/" >}}). Some relevant
metrics include:

- tetragon_events_total
- tetragon_observer_ringbuf_events_lost_total
- tetragon_map_in_use_gauge

> - tetragon_bpf_missed_events_total

Reviewer 2: fmt nit:
```suggestion
- `tetragon_events_total`
- `tetragon_observer_ringbuf_events_lost_total`
- `tetragon_map_in_use_gauge`
- `tetragon_bpf_missed_events_total`
```

The event throttling feature offers protection against event floods. This
feature monitors event generation rates per cgroup and automatically restricts
further event posting when configurable thresholds are surpassed. Upon

> activation, the system generates a THROTTLE_START event to inform operators of

Reviewer 2: ```suggestion
activation, the system generates a `THROTTLE_START` event to inform operators of
```

caption="Limited Privilege Host Attacker Diagram"
width=800px
alt="Limited privilege host attacker illustration"
>}}

A limited-privilege host attacker can escalate beyond container boundaries and
access host namespaces and resources, but lacks the root privileges required to
manipulate kernel state or disable system services. Such scenarios may result
from privilege escalation paths, including container escapes that yield
non-privileged host access, compromise of services with elevated Linux
capabilities, or exploitation of vulnerabilities that grant specific privileged
operations without full root access.

The threats posed by a limited-privilege host attacker depend on the specific
privileges and access they obtain.

- If the attacker can access the Tetragon Unix socket file at
  `/var/run/tetragon/tetragon.sock`, they can interact with Tetragon's gRPC API,
enabling them to read and potentially modify current tracing policies. Such
access undermines Tetragon's integrity by allowing the attacker to disable
monitoring, adjust policies to evade detection, or inject policies that generate
misleading events to confuse security teams. The default socket permissions of

> 0660 restrict access to the user and group running the Tetragon agent, assuming

Reviewer 2: ```suggestion
`0660` restrict access to the user and group running the Tetragon agent, assuming
```

no other processes share those credentials.
- Deployments that expose the gRPC API on a network port, rather than a Unix
  socket, introduce additional risk. An attacker with network access to the

> node's localhost interface could connect to this port and obtain the same

Reviewer 2: ```suggestion
node's `localhost` interface could connect to this port and obtain the same
```

production environments unless robust authentication is implemented.
- In Kubernetes environments, policy manipulation can have consequences that
  extend beyond the initially compromised node. Since tracing policies may be
distributed across the cluster via Custom Resource Definitions (CRDs), an
attacker who can modify policies could affect monitoring and enforcement across
the entire cluster. 
- Access to Tetragon's event export files is a significant threat vector for
  limited-privilege attackers. If an attacker can read the JSON files where
Tetragon records security events, they obtain comprehensive insight into the
security monitoring capabilities deployed on the node. This information reveals
which activities are monitored, which processes are tracked, which file accesses
are logged, and which network connections trigger alerts. With this knowledge,
an attacker can adapt their behavior to avoid detection by selecting techniques
that fall outside monitored activity patterns.

Tetragon's current design does not offer comprehensive protection against all
forms of kernel-level exploitation. For example, an attacker who acquires the

> CAP_SYS_BPF capability can load arbitrary eBPF programs into the kernel.

Reviewer 2: ```suggestion
`CAP_SYS_BPF` capability can load arbitrary eBPF programs into the kernel.
```

demonstrates that verifier bypasses are possible, potentially enabling malicious
programs to compromise kernel integrity. Tetragon cannot prevent or mitigate
such attacks when they succeed because the attacker operates at the same kernel
level as Tetragon's monitoring infrastructure. This limitation is an inherent
constraint of eBPF-based security tools within the standard Linux security
model.

**Recommended controls:**

- Protective measures against limited-privilege host attackers should prioritize
  strong access controls and minimize the attack surface. 
- Containers must not run with the same user and group IDs as the Tetragon
  agent, thereby preventing access to the Unix socket through credential
matching. The default Unix socket permissions should remain unchanged. 
- Users using network-based gRPC access should critically assess the need for
  this configuration and, if feasible, [transition to Unix socket-based access
with appropriate filesystem permissions]({{< ref
"/docs/reference/daemon-configuration/#restrict-grpc-api-access" >}}).
- Event log files require similarly strict permission management. The

>   0600(rw-------) permissions appropriately restrict access, and operators

Reviewer 2: ```suggestion
  `0600(rw-------)` permissions appropriately restrict access, and operators
```

is to stream events over gRPC with appropriate access controls, rather than
relying on lax filesystem permissions.

## Kubernetes API Server Attacker

{{< figure src="/images/threat-model/kubernetes-api.png"
caption="Kubernetes API Server Attacker Diagram" width=800px
alt="Kubernetes API Server Attacker illustration"
>}}

An attacker who compromises the Kubernetes API server or acquires credentials
for API access obtains substantial control over Tetragon's operations in
Kubernetes-based deployments. The API server serves as the control plane for
policy distribution, establishing a critical trust boundary within Tetragon's
security model. As a result:

- An attacker with read access to Tetragon's CRDs can view the TracingPolicy
  deployed throughout the cluster. This level of visibility exposes the entire
monitoring and enforcement posture, including the specific activities Tetragon
observes, the policies enforcing security boundaries, and potential gaps in
coverage. Knowledge of deployed policies enables sophisticated attackers to
develop evasion strategies by selecting attack techniques that are not monitored
or by exploiting edge cases in policy logic.

> - Write access to policy CustomResourceDefinitions constitutes a significantly

Reviewer 2: If you were talking about the Kubernetes object CustomResourceDefinitions (the k8s object that defines the tetragon custom resources) then yes but you are just saying the custom resources here.
```suggestion
- Write access to policy custom resource constitutes a significantly
```

  more severe compromise. An attacker with the ability to create, modify, or
delete TracingPolicy can fundamentally alter Tetragon's behavior across the
cluster. Such an attacker could remove existing policies to eliminate
monitoring, create permissive policies that allow previously blocked operations,
or introduce policies that generate false positive events to overwhelm security
teams and obscure genuine attacks.

**Recommended controls:**

- The primary defense against API server attacks is the rigorous implementation
  of Kubernetes Role-Based Access Control (RBAC). RBAC policies should adhere to
the principle of least privilege, granting access to Tetragon

> CustomResourceDefinitions only to users and service accounts with legitimate

Reviewer 2: ```suggestion
custom resource only to users and service accounts with legitimate
```

detection rules.

## Conclusion

While eBPF provides unparalleled visibility and enforcement capabilities, its
security is still tied to the integrity of  the adjacent environment it operates
in, particularly the host operating system and the Kubernetes control plane.
Implementing mitigation strategies that encompass the recommended controls in
this threat model will ensure Tetragon continues to remain a trusted source of
truth. Operators should read recommendations provided in this threat model and
assess which potential controls are relevant for them.

## Feedback and Responsible Disclosure

As Tetragon evolves, new features and capabilities may expand the attack surface
or alter the threat landscape. Users should remain informed about Tetragon
security updates, engage with the community to exchange threat intelligence and
defensive strategies, and periodically reassess their deployments in light of
emerging threats.  The Tetragon community encourages feedback on this threat
model and invites users to report any threats not addressed in this analysis.
Security concerns should be communicated to the private security mailing list at

> [security@cilium.io](mailto:security@cilium.io) in accordance with responsible

Reviewer 2: just a side note: maybe soon we can update to the one for `tetragon.io`, it's not yet there!

Author: Yes! I think this is an ongoing convo to create a dedicated one for tetragon.io
